### DISCUSSION TITLE
Question for QUIC Feature with preferred address migration

### DISCUSSION THREAD

**Discussion by YifangZhang** [aiortc/aioquic#503] [Q&A] [answered]
Dear authors,

I am a relatively new user of both QUIC and aioquic. 

Based on RFC 9000 [Section 9.6](https://datatracker.ietf.org/doc/html/rfc9000#section-9.6), there is a possible migration on the server side to migrate the whole connection into the server's preferred address. 

Is it possible for someone could point out this part of the code in aioquic for me? 

I have searched in the aioquic package and can only find keyword "preferred_address" in packet.py ([here](https://github.com/aiortc/aioquic/blob/a6d91fbcb5d7660adb1597bb711f6f963c598e03/src/aioquic/quic/packet.py#L284)) and once in connection.py ([here](https://github.com/aiortc/aioquic/blob/a6d91fbcb5d7660adb1597bb711f6f963c598e03/src/aioquic/quic/connection.py#L2560)). 

None of the above seems to be migration with my limited understanding. Please correct me if I am wrong on this.

Thanks a lot!

**Comment by rthalley** [comment]
As far as I can tell from the source, we currently have support for encoding and decoding, and we check that if it is present in the transport parameters that we are not a server.  I don't see any way (if we are a server) to cause it to be transmitted, nor do I see the client preserving the information and delivering it as an event, which I presume would be needed if the client application were going to consider taking the advice.  So to summarize: if we see it we can parse it but we ignore it unless we're a server in which case it is an error; we don't send it.  I defer to Jeremy if I have missed anything!

**Reply by YifangZhang** [reply]
Please correct me if I misunderstand the situation @rthalley. So, the package has supported the encoding and decoding for server-side migration but is ignored if we are a server. 

I still having trouble understanding the summary part: "if we see it we can parse it but we ignore it unless we're a server in which case it is an error; we don't send it."

On RFC9000 [section 9.6](https://datatracker.ietf.org/doc/html/rfc9000#section-9.6) says: "QUIC allows servers to accept connections on one IP address and attempt to transfer these connections to a more preferred address shortly after the handshake. This is particularly useful when clients initially connect to an address shared by multiple servers but would prefer to use a unicast address to ensure connection stability."

In my limited knowledge, I think it means the server can initiate the server-sided migration from the current address to a more preferred address (since I am new to this protocol, maybe my understanding is incorrect). So, I thought if we are a server, we can initiate the server-side migration after handshake to a new preferred address. However, I was not able to find such implementation in the aioquic package. 

And apologize ahead if this is just my misunderstanding, and thank you for the deter and answers!

**Reply by rthalley** [reply, marked as answer]
Your interpretation of the spec is correct, we just don't implement it yet.

**Comment by YifangZhang** [comment]
Hi @rthalley, I have dived deeper in the Quic's SPA, and got one addiitonal question regarding to preferred address.
Why does the constructor of QuicPreferredAddress uses:

```
class QuicPreferredAddress:
    ipv4_address: Optional[tuple[str, int]]
    ipv6_address: Optional[tuple[str, int]]
    connection_id: bytes
    stateless_reset_token: bytes
```

instead of the following:

```
class QuicPreferredAddress:
    ipv4_address: Optional[tuple[str, int]]
    ipv6_address: Optional[tuple[str, int]]
    connection_id: QuicConnectionId
```

I found that there is no sequence number passed when using the original preferred address constructor, which could result in the loss of critical information on the client side during SPA.

Please correct me if I misunderstand it.
